tests/bin/difference.py

```diff
@@ -0,0 +1,420 @@
+        raise UserWarning("Reports differ: compare with 'diff -r {reports1} {reports2}'") from error
+
+################################################################
+# Validate command line arguments
+
+def check_git_status(repo):
+    try:
+        status = run(["git", "status", "--porcelain"], cwd=repo, capture_output=True)
+    except subprocess.CalledProcessError as error:
+        raise UserWarning("This is not a git repository: git status failed") from error
+    if status:
+        raise UserWarning("This git repository has uncomitted changes.")
+    return True
+
+def validate_litani(litani):
+    logging.debug("--litani = %s", litani)
+    try:
+        return run(["which", litani], capture_output=True)[0]
+    except subprocess.CalledProcessError as error:
+        raise UserWarning("Use --litani to give the command invoking litani") from error
+
+def validate_repository(repo):
+    logging.debug("--viewer-repository = %s", repo)
+    try:
+        assert isinstance(repo, str)
+        assert os.path.exists(f"{repo}{os.sep}.git")
+        return repo
+    except AssertionError as error:
+        raise UserWarning(
+            "Use --viewer-repository to give the root of the cbmc-viewer repository"
+        ) from error
+
+def validate_proofs(proofs):
+    logging.debug("--proofs = %s", proofs)
+    try:
+        assert isinstance(proofs, str)
+        assert os.path.exists(f"{proofs}{os.sep}.litani_cache_dir")
+        return proofs
+    except AssertionError as error:
+        raise UserWarning("Use --proofs to specify a proof directory "
+                          "(containing .litani_cache_dir)") from error
+
+def validate_commits(commits):
+    logging.debug("--commits = %s", commits)
+    try:
+        assert isinstance(commits, list)
+        assert len(commits) == 2
+        return commits
+    except AssertionError as error:
+        raise UserWarning("Use --commits to specify two commits or branches to compare") from error
+
+def validate_reports(reports, commits, force):
+    logging.debug("--reports = %s", reports)
+    reports = reports or []
+    assert isinstance(reports, list)
+
+    if len(reports) == 0:
+        assert isinstance(commits, list) and len(commits) == 2
+        reports = [f'/tmp/reports/{commits[0]}', f'/tmp/reports/{commits[1]}']
+    if len(reports) != 2:
+        raise UserWarning("Use --reports to specify two report directories.")
+    if reports[0] == reports[1]:
+        reports = [f"{reports[0]}1", f"{reports[0]}2"]
+
+    if os.path.exists(reports[0]) and not force:
+        raise UserWarning(f"Path {reports[0]} exists: use --force to overwrite")
+    if os.path.exists(reports[1]) and not force:
+        raise UserWarning(f"Path {reports[1]} exists: use --force to overwrite")
+
+    return reports
+
+def validate_installs(installs, commits, force):
+    logging.debug("--installs = %s", installs)
+    installs = installs or []
+    assert isinstance(installs, list)
```

Comment:
You're not catching `AssertionError` in this method. In general, I would suggest to make `argparse` work as hard as possible for you, so `type=list` and `default=[]` in the argparser would avoid these two lines and also print a more helpful message if it goes wrong.

Reply:
Validation code made more precise and simpler.